File: src/sc_neurocore/accel/go/services/chandelier_neuron.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for chandelier_neuron

package services

import (
	"errors"
	"math"
)

const chandelierExpMax = 709.0
const chandelierExpMin = -745.0

// ChandelierNeuronState holds the neuron state
type ChandelierNeuronState struct {
	V          float64
	H          float64
	N          float64
	D          float64
	P          float64
	GNa        float64
	GK         float64
	GKv1       float64
	GKv3       float64
	GL         float64
	ENa        float64
	EK         float64
	EL         float64
	CM         float64
	Phi        float64
	Dt         float64
	VThreshold float64
}

// NewChandelierNeuron creates a new ChandelierNeuron neuron with default parameters
func NewChandelierNeuron() *ChandelierNeuronState {
	return &ChandelierNeuronState{
		V:          -65.0,
		H:          0.8,
		N:          0.1,
		D:          0.0,
		P:          0.0,
		GNa:        35.0,
		GK:         9.0,
		GKv1:       3.0,
		GKv3:       4.0,
		GL:         0.1,
		ENa:        55.0,
		EK:         -90.0,
		EL:         -65.0,
		CM:         1.0,
		Phi:        5.0,
		Dt:         0.01,
		VThreshold: -20.0,
	}
}

func chandelierFinite(x float64) bool {
	return !math.IsNaN(x) && !math.IsInf(x, 0)
}

func chandelierProbability(x float64) bool {
	return chandelierFinite(x) && x >= 0.0 && x <= 1.0
}

func checkedChandelierExp(x float64) (float64, error) {
	if !chandelierFinite(x) || x > chandelierExpMax {
		return math.NaN(), errors.New("unstable chandelier exponential argument")
	}
	if x < chandelierExpMin {
		return 0.0, nil
	}
	return math.Exp(x), nil
}

func chandelierSafeRate(a, vhalf, v, k, fallback float64) (float64, error) {
	d := v + vhalf
	if math.Abs(d) < 1e-7 {
		return fallback, nil
	}
	expTerm, err := checkedChandelierExp(-d / k)
	if err != nil {
		return math.NaN(), err
	}
	rate := a * d / (1.0 - expTerm)
	if !chandelierFinite(rate) {
		return math.NaN(), errors.New("non-finite chandelier rate candidate")
	}
	return rate, nil
}

// ValidateChandelierNeuron checks the model's physical state contract.
func ValidateChandelierNeuron(s *ChandelierNeuronState) bool {
	if s == nil {
		return false
	}
	for _, value := range []float64{s.V, s.ENa, s.EK, s.EL, s.VThreshold} {
		if !chandelierFinite(value) {
			return false
		}
	}
	for _, value := range []float64{s.H, s.N, s.D, s.P} {
		if !chandelierProbability(value) {
			return false
		}
	}
	for _, value := range []float64{s.GNa, s.GK, s.GKv1, s.GKv3, s.GL} {
		if !chandelierFinite(value) || value < 0.0 {
			return false
		}
	}
	for _, value := range []float64{s.CM, s.Phi, s.Dt} {
		if !chandelierFinite(value) || value <= 0.0 {
			return false
		}
	}
	return true
}

// Step advances the neuron by one timestep.
func (s *ChandelierNeuronState) Step(iExt float64) (int, error) {
	if !ValidateChandelierNeuron(s) || !chandelierFinite(iExt) {
		return 0, errors.New("invalid chandelier state or input")
	}
	vPrev := s.V
	nSub := int(0.5 / math.Max(s.Dt, 0.001))
	if nSub < 1 {
		nSub = 1
	}
	v, h, n := s.V, s.H, s.N
	dGate, pGate := s.D, s.P
	for i := 0; i < nSub; i++ {
		am, err := chandelierSafeRate(0.1, 35.0, v, 10.0, 1.0)
		if err != nil {
			return 0, err
		}
		bmExp, err := checkedChandelierExp(-(v + 60.0) / 18.0)
		if err != nil {
			return 0, err
		}
		mInf := am / (am + 4.0*bmExp)
		ahExp, err := checkedChandelierExp(-(v + 58.0) / 20.0)
		if err != nil {
			return 0, err
		}
		ah := 0.07 * ahExp
		bhExp, err := checkedChandelierExp(-(v + 28.0) / 10.0)
		if err != nil {
			return 0, err
		}
		bh := 1.0 / (1.0 + bhExp)
		an, err := chandelierSafeRate(0.01, 34.0, v, 10.0, 0.1)
		if err != nil {
			return 0, err
		}
		bnExp, err := checkedChandelierExp(-(v + 44.0) / 80.0)
		if err != nil {
			return 0, err
		}
		bn := 0.125 * bnExp
		hNext := h + s.Phi*(ah*(1.0-h)-bh*h)*s.Dt
		nNext := n + s.Phi*(an*(1.0-n)-bn*n)*s.Dt
		dInfExp, err := checkedChandelierExp(-(v + 50.0) / 10.0)
		if err != nil {
			return 0, err
		}
		dInf := 1.0 / (1.0 + dInfExp)
		dNext := dGate + (dInf-dGate)/150.0*s.Dt
		pInfExp, err := checkedChandelierExp(-(v + 10.0) / 10.0)
		if err != nil {
			return 0, err
		}
		pInf := 1.0 / (1.0 + pInfExp)
		pNext := pGate + s.Phi*(pInf-pGate)*s.Dt
		iNa := s.GNa * math.Pow(mInf, 3.0) * hNext * (v - s.ENa)
		iK := s.GK * math.Pow(nNext, 4.0) * (v - s.EK)
		iKv1 := s.GKv1 * math.Pow(dNext, 4.0) * (v - s.EK)
		iKv3 := s.GKv3 * pNext * (v - s.EK)
		iL := s.GL * (v - s.EL)
		vNext := v + (-iNa-iK-iKv1-iKv3-iL+iExt)/s.CM*s.Dt
		if !(chandelierFinite(vNext) && vNext >= -100.0 && vNext <= 60.0 && chandelierProbability(hNext) && chandelierProbability(nNext) && chandelierProbability(dNext) && chandelierProbability(pNext)) {
			return 0, errors.New("invalid chandelier candidate state")
		}
		v, h, n = vNext, hNext, nNext
		dGate, pGate = dNext, pNext
	}
	s.V, s.H, s.N = v, h, n
	s.D, s.P = dGate, pGate
	if s.V >= s.VThreshold && vPrev < s.VThreshold {
		return 1, nil
	}
	return 0, nil
}

// SimulateChandelierNeuron runs the neuron for n steps.
func SimulateChandelierNeuron(nSteps int, iExt float64) ([]float64, int) {
	if nSteps < 0 {
		return nil, 0
	}
	s := NewChandelierNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		result, err := s.Step(iExt)
		if err != nil {
			trace[t] = math.NaN()
			continue
		}
		trace[t] = s.V
		if result > 0 {
			spikes++
		}
	}
	return trace, spikes
}

File: src/sc_neurocore/accel/julia/neurons/chandelier_neuron.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for chandelier_neuron

module ChandelierNeuronAccel

export step!, simulate, validate, ChandelierNeuronState

mutable struct ChandelierNeuronState
    v::Float64
    h::Float64
    n::Float64
    d::Float64
    p::Float64
    g_na::Float64
    g_k::Float64
    g_kv1::Float64
    g_kv3::Float64
    g_l::Float64
    e_na::Float64
    e_k::Float64
    e_l::Float64
    c_m::Float64
    phi::Float64
    dt::Float64
    v_threshold::Float64
end

function ChandelierNeuronState()
    ChandelierNeuronState(-65.0, 0.8, 0.1, 0.0, 0.0, 35.0, 9.0, 3.0, 4.0, 0.1, 55.0, -90.0, -65.0, 1.0, 5.0, 0.01, -20.0)
end

_probability(x::Float64)::Bool = isfinite(x) && 0.0 <= x <= 1.0

function validate(s::ChandelierNeuronState)::Bool
    all(isfinite, (s.v, s.e_na, s.e_k, s.e_l, s.v_threshold)) &&
        all(_probability, (s.h, s.n, s.d, s.p)) &&
        all(x -> isfinite(x) && x >= 0.0, (s.g_na, s.g_k, s.g_kv1, s.g_kv3, s.g_l)) &&
        all(x -> isfinite(x) && x > 0.0, (s.c_m, s.phi, s.dt))
end

function _checked_exp(x::Float64)::Float64
    if !isfinite(x) || x > 709.0
        return NaN
    elseif x < -745.0
        return 0.0
    end
    exp(x)
end

function _safe_rate(a::Float64, vhalf::Float64, v::Float64, k::Float64, fallback::Float64)::Float64
    d = v + vhalf
    if abs(d) < 1e-7
        return fallback
    end
    rate = a * d / (1.0 - _checked_exp(-d / k))
    isfinite(rate) ? rate : NaN
end

function step!(s::ChandelierNeuronState, I_ext::Float64=0.0; dt::Float64=0.1)
    if !validate(s) || !isfinite(I_ext)
        return -1
    end
    v_prev = s.v
    n_sub = max(1, Int(0.5 / max(s.dt, 0.001)))
    v, h, n, d_gate, p_gate = s.v, s.h, s.n, s.d, s.p
    for _ in 1:n_sub
        am = _safe_rate(0.1, 35.0, v, 10.0, 1.0)
        bm = 4.0 * _checked_exp(-(v + 60.0) / 18.0)
        m_inf = am / (am + bm)
        ah = 0.07 * _checked_exp(-(v + 58.0) / 20.0)
        bh = 1.0 / (1.0 + _checked_exp(-(v + 28.0) / 10.0))
        an = _safe_rate(0.01, 34.0, v, 10.0, 0.1)
        bn = 0.125 * _checked_exp(-(v + 44.0) / 80.0)
        h_next = h + s.phi * (ah * (1.0 - h) - bh * h) * s.dt
        n_next = n + s.phi * (an * (1.0 - n) - bn * n) * s.dt
        d_inf = 1.0 / (1.0 + _checked_exp(-(v + 50.0) / 10.0))
        d_next = d_gate + (d_inf - d_gate) / 150.0 * s.dt
        p_inf = 1.0 / (1.0 + _checked_exp(-(v + 10.0) / 10.0))
        p_next = p_gate + s.phi * (p_inf - p_gate) * s.dt
        i_na = s.g_na * m_inf ^ 3 * h_next * (v - s.e_na)
        i_k = s.g_k * n_next ^ 4 * (v - s.e_k)
        i_kv1 = s.g_kv1 * d_next ^ 4 * (v - s.e_k)
        i_kv3 = s.g_kv3 * p_next * (v - s.e_k)
        i_l = s.g_l * (v - s.e_l)
        v_next = v + (-i_na - i_k - i_kv1 - i_kv3 - i_l + I_ext) / s.c_m * s.dt
        if !(isfinite(v_next) && -100.0 <= v_next <= 60.0 && all(_probability, (h_next, n_next, d_next, p_next)))
            return -1
        end
        v, h, n, d_gate, p_gate = v_next, h_next, n_next, d_next, p_next
    end
    s.v, s.h, s.n, s.d, s.p = v, h, n, d_gate, p_gate
    return (s.v >= s.v_threshold && v_prev < s.v_threshold) ? 1 : 0
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    s = ChandelierNeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext; dt=dt)
        trace[t] = s.v
        if result isa Number && result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module ChandelierNeuronAccel
